### PR TITLE
Convert Dock class to Etch

### PR DESCRIPTION
**Pull request by matthewwithanm**
Since #16116 is going to require big changes to how docks are layed out (for which I'll be using etch), I wanted to convert Dock itself. Then I'll update `Dock::render()` to replace the PaneContainer with a `DockGroupContainer` (which itself will render `DockGroups`).

Dock was already written in a super Reacty way so there isn't much being changed here. Basically, `render()` just returns vdom instead of being a side-effect cesspool.

While working on this, I (re)discovered #16769. In order to be able to verify that I didn't make a mess of things with this PR, I fixed that issue in #16863 and rebased on that. I opened this as a PR against that branch for review purposes, but after that's merged I'll rebase it onto master and switch the base. Please let me know if there was a better way to do it!

### CODE REVIEW

**Comment by matthewwithanm** [comment]
it's choking on the jsx when building the snapshot [here](https://github.com/eventualbuddha/ast-util/blob/6ea220d1999979c970c6a0a11faf4c9db4c1fddf/lib/index.js#L468)…does this ring any bells for you guys? i'm not familiar with this part of the codebase 😞

**Comment by lee-dohm** [comment]
#16936 should be merged at the same time as this.

**Review by daviwil — APPROVED** [review]
Looks great to me!

**Comment by thomasjo** [comment]
Looks like this broke something. Both @lee-dohm and myself are unable to successfully launch Atom when building from source. Judging from the comments in the code related to snapshotting, I'm guessing something is failing there. Seeing errors like this (in safe mode) upon launching Atom:

```
Uncaught TypeError: parentElement.insertBefore is not a function
    at updateChildren (/Applications/Atom.app/Contents/Resources/app/node_modules/etch/lib/patch.js:123:27)
    at patch (/Applications/Atom.app/Contents/Resources/app/node_modules/etch/lib/patch.js:18:15)
    at updateSync (/Applications/Atom.app/Contents/Resources/app/node_modules/etch/lib/component-helpers.js:116:26)
    at /Applications/Atom.app/Contents/Resources/app/node_modules/etch/lib/component-helpers.js:72:13
    at ViewRegistry.performDocumentUpdate (/Applications/Atom.app/Contents/Resources/app/src/view-registry.js:238:13)
<embedded>:138043 Uncaught (in promise) TypeError: parentElement.insertBefore is not a function
    at updateChildren (/Applications/Atom.app/Contents/Resources/app/node_modules/etch/lib/patch.js:123:27)
    at patch (/Applications/Atom.app/Contents/Resources/app/node_modules/etch/lib/patch.js:18:15)
    at Object.updateSync (/Applications/Atom.app/Contents/Resources/app/node_modules/etch/lib/component-helpers.js:116:26)
    at Dock.getElement (/Applications/Atom.app/Contents/Resources/app/src/dock.js:88:18)
    at HTMLElement.initialize (/Applications/Atom.app/Contents/Resources/app.asar/src/panel-container-element.js:27:40)
    at PanelContainer.getElement (/Applications/Atom.app/Contents/Resources/app/src/panel-container.js:29:64)
    at HTMLElement.initialize (/Applications/Atom.app/Contents/Resources/app.asar/src/workspace-element.js:117:45)
    at Workspace.getElement (/Applications/Atom.app/Contents/Resources/app/src/workspace.js:254:59)
    at AtomEnvironment.registerDefaultTargetForKeymaps (/Applications/Atom.app/Contents/Resources/app/src/atom-environment.js:339:55)
    at loadStatePromise.loadState.then (/Applications/Atom.app/Contents/Resources/app/src/atom-environment.js:831:18)
    at <anonymous>
```

@matthewwithanm Any ideas? Only workaround we've got so far is reverting to a commit before this got merged.

**Comment by matthewwithanm** [comment]
weird, i'll try to repro.

**Comment by thomasjo** [comment]
@matthewwithanm Let me know if you need anything else, but this is trivially reproducible merely by building from `master`. We're at least three maintainers so far that are hitting this. Note that we've all setup Atom for full local development, with `$ATOM_DEV_RESOURCE_PATH` configured, and so on.

My hunch is that this might somehow be related to snapshotting and/or ASAR packaging, since that is the only difference I can think of that sets CI builds apart from local development builds. Not sure if that makes sense though 🤔

**Comment by matthewwithanm** [comment]
👍waiting for a rebuild now. i've been running from source and it wasn't reproing.

**Comment by matthewwithanm** [comment]
finally finished cleaning and rebuilding and…no repro. any ideas?

**Comment by Arcanemagus** [comment]
Are you running in `--dev`? I only see this when running the build as a "normal" Atom (no `--dev`, no `ATOM_DEV_RESOURCE_PATH`).

**Comment by lee-dohm** [comment]
@Arcanemagus No, I wasn't running in Dev Mode when I encountered this error.

**Comment by thomasjo** [comment]
Curiously, running in dev mode (`--dev`) makes the problem go away.

**Comment by lee-dohm** [comment]
Just ran a test using Atom v1.27.0-dev-a066295c7 on Mac OS X 10.13.4 after:

1. Execute `script/clean`
1. Execute `git clean -f`
1. Execute `script/build --install`
1. Execute `atom --dev .` (everything works)
1. Execute `atom .`

I get the same problems as before.

**Comment by daviwil** [comment]
Hey @matthewwithanm, since these changes are causing issues for folks on master, I'm considering reverting those two PRs until we figure out the problem.  Do you mind if I revert https://github.com/atom/atom/commit/31e9ddad1e9a1d9e9cd69e3500e8729afa585f85 and https://github.com/atom/atom/commit/a066295c728ca79fbc91057836fc0ed5517c5111 in master so that we can continue the investigation in a separate branch?

**Comment by matthewwithanm** [comment]
ah! i was running in dev! thanks for the pointer!

@daviwil yeah, guess we gotta 😞

**Pull request merged**
